packages/utils/src/lib/sort-report.ts

```diff
@@ -0,0 +1,77 @@
+import { CategoryRef } from '@code-pushup/models';
+import {
+  compareIssues,
+  getAuditByRef,
+  getGroupWithAudits,
+  sortAudits,
+  sortCategoryAudits,
+} from './report';
+import {
+  EnrichedScoredAuditGroupWithAudits,
+  ScoredCategoryConfig,
+  ScoredReport,
+  WeighedAuditReport,
+} from './scoring';
+
+export function sortReport(report: ScoredReport): ScoredReport {
+  const { categories, plugins } = report;
+  const sortedCategories = categories.reduce<ScoredCategoryConfig[]>(
+    (acc, category) => {
+      const { audits, groups } = category.refs.reduce(
+        (
+          acc: {
+            audits: WeighedAuditReport[];
+            groups: EnrichedScoredAuditGroupWithAudits[];
+          },
+          ref: CategoryRef,
+        ) => ({
+          ...acc,
+          ...(ref.type === 'group'
+            ? {
+                groups: [
+                  ...acc.groups,
+                  getGroupWithAudits(ref.slug, ref.plugin, plugins),
+                ],
+              }
+            : {
+                audits: [...acc.audits, getAuditByRef(ref, plugins)],
+              }),
+        }),
+        { groups: [], audits: [] },
+      );
+      const sortedAuditsAndGroups = [
+        ...groups,
+        ...audits.sort(sortCategoryAudits),
+      ];
+      const sortedRefs = category.refs.slice().sort((a, b) => {
+        const aIndex = sortedAuditsAndGroups.findIndex(
+          ref => ref.slug === a.slug,
+        );
+        const bIndex = sortedAuditsAndGroups.findIndex(
+          ref => ref.slug === b.slug,
+        );
+        return aIndex - bIndex;
+      });
+
+      return [...acc, { ...category, refs: sortedRefs }];
+    },
+    [],
+  );
```

Comment:
@MishaSeredenkoPushBased I notice you have a tendency to overuse `reduce`, when a `map`, `filter`, `flatMap`, `join` or some combination of them could be applied instead.

Take a step back and notice the pattern:

```ts
categories.reduce<ScoredCategoryConfig>(
  (acc, category) => {
    // ...
    return [...acc, { ...category, refs: sortedRefs }]
  },
  []
);
```

:top: The accumulator is an array, and each new accumulator is a copy of the previous array with an additional item at the end, which is derived from the current item. In other words, you've reinvented `Array.map`:

```ts
categories.map((category): ScoredCategoryConfig => {
  // ...
  return { ...category, refs: sortedRefs };
});
```

The `reduce` also works of course, but it's not the most appropriate method here. Using a method like `map` is a clearer statement of intent ("I'm aggregating something" :arrow_right: "I'm transforming each item"). While `reduce` is more flexible, it's also more cumbersome - more noise, explicit type annotations needed (because the accumulator could be anything), and nested `reduce`s are horrible to read (2 accumulator variables - either shadowed or have awkward names).

So I would discourage you from reaching for `reduce` as a goto solution whenever you need to loop over an array, it's more of a last resort if none of the other specialized methods fit. A good rule of thumb is if your accumulator is an array, then there's probably a better method. Some examples:
- `arr.reduce((acc, item) => [...acc, f(item)], [])` :arrow_right: `arr.map(f)`
- `arr.reduce((acc, item) => f(item) ? acc : [...acc, item], [])` :arrow_right: `arr.filter(f)`
- `arr.reduce((acc, item) => f1(item) ? acc : [...acc, f2(item)], [])` :arrow_right: `arr.filter(f1).map(f2)`
- `arr.reduce((acc, item) => [...acc, ...f(item)], [])` :arrow_right: `arr.flatMap(f)`

Reply:
Thanks for the great explanation! ❤️  I've used `map` instead of outer `reduce` in my case.